alf/utils/lean_function.py

```diff
@@ -13,67 +13,123 @@
+        # it can get the information about the format of the output. For other
+        # types of func, if ret is not a Tensor or tuple of Tensors, pytorch will
+        # report an error.
+        if isinstance(func, Network):
+            ret = tuple(flatten(ret))
+        return ret
 
     @staticmethod
     def backward(ctx, grad_output):
         with torch.enable_grad():
-            tensors = [t.clone() for t in ctx.saved_tensors]
+            # saved_tensors is the tensors passed for ctx.save_for_backward
+            tensors = list(ctx.saved_tensors)
             func = ctx.func
+            parameters = ctx.parameters
+            num_parameters = len(parameters)
             args = tuple(
                 tensors.pop(0) if arg[0] else arg[1] for arg in ctx.args)
             tensors = tuple(arg for i, arg in enumerate(args)
-                            if ctx.needs_input_grad[i + 2])
+                            if ctx.needs_input_grad[3 + num_parameters + i])
             keywords = ctx.keywords
+            func._inside_lean_function = True
             if keywords:
                 num_kwargs = len(keywords)
                 kwargs = dict(zip(keywords, args[-num_kwargs:]))
                 args = args[:-num_kwargs]
                 out = func(*args, **kwargs)
             else:
                 out = func(*args)
-        grads = list(torch.autograd.grad(out, tensors, grad_output))
+            func._inside_lean_function = False
+        if isinstance(func, Network):
+            out = tuple(flatten(out))
+        grads = list(
+            torch.autograd.grad(out, parameters + tensors, grad_output))
         grads = tuple(
             grads.pop(0) if need else None for need in ctx.needs_input_grad)
         return grads
 
 
-def lean_function(func):
+def lean_function(func: Callable) -> Callable:
     """Wrap ``func`` to save memory for backward.
 
     The returned function performs same computation as ``func``, but save memory
     by discarding intermediate results. It calculates the gradient by recomputing
     ``func`` using the same input during backward.
 
-    Note: All the Tensor inputs to ``func`` must be explicitly listed as arguments
-    of ``func``. For example, a tuple of Tensors as argument is not allowed.
-    Another requirement for ``func`` is that it must be deterministic so that
-    repeated evaluation with the same input will get same output. It is the
-    responsibility of the user of this function to make sure that ``func``
-    satifisies these requirements. ``lean_function`` will not report error if
-    ``func`` does not satisfies these requirements and error will be silently
-    ignored.
+    Note: There are several requirements for ``func``:
+
+        1. All the Tensor inputs to ``func`` must be explicitly listed as arguments
+          of ``func``. For example, a tuple of Tensors as argument is not allowed.
+          Using Tensors outside of ``func`` is not allowed either unless ``func``
```

Comment:
Maybe explicitly explain that, for `nn.Module` we do not have to put the parameters in the argument list because it will be automatically taken care of.

Reply:
updated the comment